Apply modernize-use-nullptr on header files.

quantumsteve opened the pull request:
This uses clang-tidy's modernize-use-nullptr tool to replace NULL and 0 with nullptr. Unlike an earlier PR, this includes changes in headers.

code review: This is a relatively simple change done with automated tooling. Rather than reviewing every change I suggest reviewing enough files to convince yourself clang-tidy is doing the right things.

testing: A misspelled or duplicated nullptr should not build.

SimonHeybrock commented on Framework/API/inc/MantidAPI/ImplicitFunctionParser.h:
```diff
@@ -83,7 +83,7 @@ class MANTID_API_DLL ImplicitFunctionParser {
 
   void checkSuccessorExists() {
-    if (0 == m_successor.get()) {
+    if (nullptr == m_successor.get()) {
```
`if(!m_successor)`

SimonHeybrock commented:
Looks all good. There are a several things written as `if(ptr != nullptr)` that could just be written as `if(ptr)`, but I guess that is a matter of taste.

quantumsteve commented:
I cleaned up a few comparisons and changed a few back to make Windows happy.

I left files in the `ANN` and `GSoap` directories alone since they are rarely read by developers and I don't want to make a subtle mistake.
